src/custom_token.rs

```diff
@@ -0,0 +1,534 @@
+// Not public API, but #[macro_export] _is_ documented
+#![doc(hidden)]
+
+#[cfg(feature = "parsing")]
+use token;
+#[cfg(feature = "parsing")]
+use buffer::Cursor;
+
+#[cfg(feature = "parsing")]
+pub trait CustomToken {
+    fn peek(cursor: Cursor) -> bool;
+    fn display() -> &'static str;
+}
+
+#[cfg(feature = "parsing")]
+impl<T: CustomToken> token::private::Sealed for T {}
+#[cfg(feature = "parsing")]
+impl<T: CustomToken> token::Token for T {
+    fn peek(cursor: Cursor) -> bool {
+        <Self as CustomToken>::peek(cursor)
+    }
+
+    fn display() -> &'static str {
+        <Self as CustomToken>::display()
+    }
+}
+
+// Not public API.
+#[doc(hidden)]
+#[macro_export]
+macro_rules! punct_len {
+    ()       => ( 0usize );
+    (+)      => ( 1usize );
+    (+=)     => ( 2usize );
+    (&)      => ( 1usize );
+    (&&)     => ( 2usize );
+    (&=)     => ( 2usize );
+    (@)      => ( 1usize );
+    (!)      => ( 1usize );
+    (^)      => ( 1usize );
+    (^=)     => ( 2usize );
+    (:)      => ( 1usize );
+    (::)     => ( 2usize );
+    (,)      => ( 1usize );
+    (/)      => ( 1usize );
+    (/=)     => ( 2usize );
+    (.)      => ( 1usize );
+    (..)     => ( 2usize );
+    (...)    => ( 3usize );
+    (..=)    => ( 3usize );
+    (=)      => ( 1usize );
+    (==)     => ( 2usize );
+    (>=)     => ( 2usize );
+    (>)      => ( 1usize );
+    (<=)     => ( 2usize );
+    (<)      => ( 1usize );
+    (*=)     => ( 2usize );
+    (!=)     => ( 2usize );
+    (|)      => ( 1usize );
+    (|=)     => ( 2usize );
+    (||)     => ( 2usize );
+    (#)      => ( 1usize );
+    (?)      => ( 1usize );
+    (->)     => ( 2usize );
+    (<-)     => ( 2usize );
+    (%)      => ( 1usize );
+    (%=)     => ( 2usize );
+    (=>)     => ( 2usize );
+    (;)      => ( 1usize );
+    (<<)     => ( 2usize );
+    (<<=)    => ( 3usize );
+    (>>)     => ( 2usize );
+    (>>=)    => ( 3usize );
+    (*)      => ( 1usize );
+    (-)      => ( 1usize );
+    (-=)     => ( 2usize );
+    (~)      => ( 1usize );
+    (_)      => ( 1usize );
+    ($tt:tt) => ( compile_error!("Punctuation `{}` is not supported", stringify!($tt)) );
+    ($head:tt $($tail:tt)+) => ( punct_len!($head) + punct_len!($($tail)+) );
```

Comment:
I would write this as follows. Nobody should be writing custom punctuation with 64+ characters but why mess with recursion when not necessary.

```rust
punct_len!($head) $(+ punct_len!($tail))*
```